onnxruntime/src/runner.rs

```diff
@@ -0,0 +1,244 @@
+}
+
+fn tensors_to_ptr<'a, 's: 'a, T: Element, D: Dimension + 'a>(
+    tensors: impl IntoIterator<Item = &'a OrtTensor<'s, T, D>>,
+) -> Vec<*const sys::OrtValue> {
+    tensors
+        .into_iter()
+        .map(|tensor| tensor.c_ptr as *const _)
+        .collect()
+}
+
+fn tensors_to_mut_ptr<'a, 's: 'a, T: Element, D: Dimension + 'a>(
+    tensors: impl IntoIterator<Item = &'a mut OrtTensor<'s, T, D>>,
+) -> Vec<*mut sys::OrtValue> {
+    tensors
+        .into_iter()
+        .map(|tensor| tensor.c_ptr as *mut _)
+        .collect()
+}
+
+fn arrays_to_ort<T: Element, D: Dimension>(
+    memory_info: &MemoryInfo,
+    arrays: impl IntoIterator<Item = Array<T, D>>,
+) -> Result<(Vec<OrtTensor<T, D>>, Vec<*const sys::OrtValue>)> {
+    let ort_tensors = arrays
+        .into_iter()
+        .map(|arr| OrtTensor::from_array(memory_info, arr))
+        .collect::<Result<Vec<_>>>()?;
+    let ort_values = ort_tensors
+        .iter()
+        .map(|tensor| tensor.c_ptr as *const _)
+        .collect();
+    Ok((ort_tensors, ort_values))
+}
+
+fn arrays_with_shapes<T: Element, D: Dimension>(shapes: &[Vec<usize>]) -> Result<Vec<Array<T, D>>> {
+    Ok(shapes
+        .into_iter()
+        .map(|shape| Array::<_, IxDyn>::default(shape.clone()).into_dimensionality())
+        .collect::<std::result::Result<Vec<Array<T, D>>, _>>()?)
+}
+
+pub struct Inputs<'r, 'a, T: Element, D: Dimension> {
+    tensors: &'a mut [OrtTensor<'r, T, D>],
+}
+
+impl<T: Element, D: Dimension> Inputs<'_, '_, T, D> {}
+
+impl<T: Element, D: Dimension> Index<usize> for Inputs<'_, '_, T, D> {
+    type Output = Array<T, D>;
+
+    #[inline]
+    fn index(&self, index: usize) -> &Self::Output {
+        &(*self.tensors[index])
+    }
+}
+
+impl<T: Element, D: Dimension> IndexMut<usize> for Inputs<'_, '_, T, D> {
+    #[inline]
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        &mut (*self.tensors[index])
+    }
+}
+
+pub struct Outputs<'r, 'a, T: Element, D: Dimension> {
+    tensors: &'a [OrtTensor<'r, T, D>],
+}
+
+impl<T: Element, D: Dimension> Outputs<'_, '_, T, D> {}
+
+impl<T: Element, D: Dimension> Index<usize> for Outputs<'_, '_, T, D> {
+    type Output = Array<T, D>;
+
+    #[inline]
+    fn index(&self, index: usize) -> &Self::Output {
+        &(*self.tensors[index])
+    }
+}
+
+pub struct RunnerBuilder<'s, TIn: Element, DIn: Dimension> {
+    session: &'s Session,
+    input_arrays: Vec<Array<TIn, DIn>>,
+}
+
+impl<'s, TIn: Element, DIn: Dimension> RunnerBuilder<'s, TIn, DIn> {
+    #[inline]
+    pub fn new(
+        session: &'s Session,
+        input_arrays: impl IntoIterator<Item = Array<TIn, DIn>>,
+    ) -> Self {
+        Self {
+            session,
+            input_arrays: input_arrays.into_iter().collect(),
+        }
+    }
+
+    #[inline]
+    pub fn with_output<TOut: Element, DOut: Dimension>(
+        self,
+    ) -> Result<Runner<'s, TIn, DIn, TOut, DOut>> {
+        Runner::new(self.session, self.input_arrays)
+    }
```

Comment:
This is meant to be used like this:
```rust
    let mut runner = session
        .make_runner(input_tensor_values)
        .with_output::<f32, Ix4>()?;
```

Since the model (and thus the session) knows the shape of the output, can this be leveraged? Or would that require const generics (probably...)?

Reply:
Ok, looking at the following `with_output_dyn()` answers the comment above, please ignore.